File: src/lib/config/feeds.ts
import { readFile } from 'node:fs/promises';
import path from 'node:path';

export type FeedFormat = 'rss' | 'rdf';

export type FeedConfig = {
  id: string;
  name: string;
  url: string;
  format: FeedFormat;
};

export type FeedTopic = {
  id: string;
  label: string;
  feeds: FeedConfig[];
};

export type FeedsConfig = {
  topics: FeedTopic[];
};

export async function loadFeedsConfig(): Promise<FeedsConfig> {
  const filePath = path.join(process.cwd(), 'feeds.json');
  const raw = await readFile(filePath, 'utf-8');
  return JSON.parse(raw) as FeedsConfig;
}

File: cmd/ingest/feeds/feeds.go
package feeds

import (
	"encoding/json"
	"fmt"
	"os"
)

type FeedConfig struct {
	ID     string `json:"id"`
	Name   string `json:"name"`
	URL    string `json:"url"`
	Format string `json:"format"`
}

type FeedTopic struct {
	ID    string       `json:"id"`
	Label string       `json:"label"`
	Feeds []FeedConfig `json:"feeds"`
}

type FeedsFile struct {
	Topics []FeedTopic `json:"topics"`
}

type FeedEntry struct {
	TopicID    string
	TopicLabel string
	Feed       FeedConfig
}

func Load(path string) ([]FeedEntry, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("reading feeds file: %w", err)
	}

	var f FeedsFile
	if err := json.Unmarshal(data, &f); err != nil {
		return nil, fmt.Errorf("parsing feeds file: %w", err)
	}

	var entries []FeedEntry
	for _, topic := range f.Topics {
		for _, feed := range topic.Feeds {
			entries = append(entries, FeedEntry{
				TopicID:    topic.ID,
				TopicLabel: topic.Label,
				Feed:       feed,
			})
		}
	}
	return entries, nil
}
